src/nvim/api/extmark.h

```diff
@@ -4,14 +4,26 @@
 
 #include "nvim/api/keysets_defs.h"  // IWYU pragma: keep
 #include "nvim/api/private/defs.h"  // IWYU pragma: keep
+#include "nvim/buffer_defs.h"
 #include "nvim/decoration_defs.h"  // IWYU pragma: keep
 #include "nvim/macros_defs.h"
 #include "nvim/map_defs.h"
 #include "nvim/types_defs.h"
 
 EXTERN Map(String, int) namespace_ids INIT( = MAP_INIT);
+EXTERN Set(uint32_t) namespace_scoped INIT( = SET_INIT);
 EXTERN handle_T next_namespace_id INIT( = 1);
 
+/// Returns true if the namespace is scoped in the given window.
+static inline bool ns_in_win(uint32_t ns_id, win_T *wp)
+{
+  if (!set_has(uint32_t, &namespace_scoped, ns_id)) {
+    return true;
+  }
+
+  return set_has(uint32_t, &wp->w_ns_set, ns_id);
```

Comment:
This seems strange. Why is `namespace_scoped` necessary? If ns_id is *not* in `namespace_scoped`, this function returns `true`? Why isn't the decision fully decided by `w_ns_set`, why is `namespace_scoped` needed at all?

Reply:
`namespace_scoped` is needed because it currently is equivalent to "namespace is scoped in at least one window" and if it doesn't contain the namespace, then that namespace is "global scoped" (which is what you asked for).

Though `namespace_scoped` can sometimes contain namespaces that are not scoped in any windows, because if a window is deleted, and that window was the only window where the namespace was scoped, then `namespace_scoped` isn't updated. (I could change this so that it's always synced, but that would entail that when a window is deleted, a namespace can suddenly become global scope, which is probably not what one wants.)

Reply:
right, the docstring and names were confusing.



> `namespace_scoped` is needed because it currently is equivalent to "namespace is scoped in at least one window"

Sounds like an unnecessary optimization. We don't need `namespace_scoped`, we can iterate through all windows and check `w_ns_set`. That is less confusing and requires less bookkeeping.